Source/Core/Core/IOS/USB/Emulated/Infinity.cpp

```diff
@@ -0,0 +1,978 @@
+{
+  std::array<u8, 16> figure_data = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                                    0x00, 0x00, 0x00, 0x01, 0xD1, 0x1F};
+  // Create block 1 of the figure, based on the figure number and type
+  InfinityFigureInfo figure = {InfinityFigureType::Character, 0};
+  for (u8 i = 0; i < list_infinity_figures.size(); i++)
+  {
+    if (list_infinity_figures[i].second.figure_number == figure_num)
+    {
+      figure = list_infinity_figures[i].second;
+      break;
+    }
+  }
+
+  switch (figure.type)
+  {
+  case InfinityFigureType::Character:
+  {
+    figure_data[1] = 0x0F;
+    figure_data[4] = 0x10;
+    figure_data[5] = 0x0C;
+    figure_data[6] = 0x1E;
+    break;
+  }
+  case InfinityFigureType::Playset:
+  {
+    figure_data[1] = 0x1E;
+    figure_data[4] = 0x11;
+    figure_data[5] = 0x04;
+    figure_data[6] = 0x06;
+    break;
+  }
+  case InfinityFigureType::Ability_A:
+  {
+    figure_data[1] = 0x2D;
+    figure_data[4] = 0x11;
+    figure_data[5] = 0x04;
+    figure_data[6] = 0x16;
+    break;
+  }
+  case InfinityFigureType::Ability_B:
+  {
+    figure_data[1] = 0x2D;
+    figure_data[4] = 0x11;
+    figure_data[5] = 0x04;
+    figure_data[6] = 0x08;
+    break;
+  }
+  case InfinityFigureType::Item:
+  {
+    figure_data[1] = 0x3D;
+    figure_data[4] = 0x12;
+    figure_data[5] = 0x0A;
+    figure_data[6] = 0x15;
+    break;
+  }
+
+  default:
+    return figure_data;
+  }
+  figure_data[2] = u8((figure_num >> 8) & 0xFF);
+  figure_data[3] = u8(figure_num & 0xFF);
+  u32 checksum = InfinityCRC32(figure_data.data(), 12);
+  for (s8 i = 3; i >= 0; i--)
+  {
+    figure_data[15 - i] = u8((checksum >> i * 8) & 0xFF);
+  }
+  DEBUG_LOG_FMT(IOS_USB, "Block 1: \n{}", HexDump(figure_data.data(), 16));
+  return figure_data;
+}
+
+void InfinityBase::GenerateSeed(u32 seed)
+{
+  m_random_a = 0xF1EA5EED;
+  m_random_b = seed;
+  m_random_c = seed;
+  m_random_d = seed;
+
+  for (int i = 0; i < 23; i++)
+  {
+    GetNext();
+  }
+}
+
+u32 InfinityBase::GetNext()
+{
+  u32 a = m_random_a;
+  u32 b = m_random_b;
+  u32 c = m_random_c;
+  u32 ret = std::rotl(m_random_b, 27);
+
+  u32 temp = (a + ((ret ^ 0xFFFFFFFF) + 1));
+  b = (b ^ std::rotl(c, 17));
+  a = m_random_d;
+  c = (c + a);
+  ret = (b + temp);
+  a = (a + temp);
+
+  m_random_c = a;
+  m_random_a = b;
+  m_random_b = c;
+  m_random_d = ret;
+
+  return ret;
+}
+
+u64 InfinityBase::Scramble(u32 num_to_scramble, u32 garbage)
+{
+  u64 mask = 0x8E55AA1B3999E8AA;
+  u64 ret = 0;
+
+  for (int i = 0; i < 64; i++)
+  {
+    ret <<= 1;
+
+    if ((mask & 1) != 0)
+    {
+      ret |= (num_to_scramble & 0x01);
+      num_to_scramble >>= 1;
+    }
+    else
+    {
+      ret |= (garbage & 0x01);
+      garbage >>= 1;
+    }
+
+    mask = mask >> 1;
+  }
+
+  return ret;
+}
+
+u32 InfinityBase::Descramble(u64 num_to_descramble)
+{
+  u64 mask = 0x8E55AA1B3999E8AA;
+  u32 ret = 0;
+
+  for (int i = 0; i < 64; i++)
+  {
+    if ((mask & (1L << 63)) > 0)
```

Comment:
This overflows on Windows, long is only 32 bits long. You may want to replace this with `Common::ExtractBit(mask, 63)` for clarity, assuming that is what you want to do here.